### PR TITLE
Modal window

### PR DESCRIPTION
**Pull request by tholex**
Simple modal that includes an X and a curtain to close it. Mounts in-place (so it has to be in a non-hidden, non-transformed div).

The demo includes a button in the center to close the modal but this is totally optional - the host application would chose what to put in the modal.

Should I add an optional width= prop?

screenshot 2015-08-24 13 36 16

### CODE REVIEW

File: src/Modal/Modal.jsx [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      "tui-modal-wrapper__hidden": (!controlledByParent) && (!isOpen)
+    });
+    const modalClasses = cx("tui-modal-content", className);
+
+    return (
+      <div className={wrapperClasses}>
+        <div className="tui-modal-curtain" onClick={closeModal}/>
+        <div className={modalClasses}>
+          <a className="tui-modal-close-button" onClick={closeModal}>
+            <span aria-hidden="true" className="icon-close" />
```
Comment (learykara) [inline src/Modal/Modal.jsx]: why not use the tui `Icon` component?

File: src/Modal/Modal.jsx [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  }
+
+  render() {
+    const {className} = this.props;
+    const {controlledByParent, isOpen} = this.state;
+
+    const closeModal = this.props.close || this._closeModal.bind(this);
+
+    const wrapperClasses = cx("tui-modal-wrapper", {
+      "tui-modal-wrapper__hidden": (!controlledByParent) && (!isOpen)
```
Comment (learykara) [inline src/Modal/Modal.jsx]: so this only needs to be hidden if it's not controlled by the parent because if it is controlled in the parent (as in HomePage), then it only exists on the page when `modalIsOpen` is `true`, correct?

File: client/routes.jsx [hunk 3]
```diff
@@ -22,7 +23,8 @@ class App extends React.Component {
 
 const routes = (
   <Route name="app" path="/" handler={App}>
-    <Route name="demo" path="demo" handler={DemoPage}/>
+    <Route name="demo" path="demo" handler={DemoPage}></Route>
```
Comment (learykara) [inline client/routes.jsx]: why this change?

File: src/Modal/Modal.jsx (hunk 2 above)
Comment (learykara) [inline src/Modal/Modal.jsx]: ah, just read the comments above. so yeah that makes sense

File: client/HomePage.jsx [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  render() {
+    const {modalIsOpen} = this.state;
+    return (<div className="tui-demo-page">
+      <h1>Thinkful UI</h1>
+      <p>Shared React components for various Thinkful apps. The nav above
+      doesn't respond to your actual login-state, but uses a fake <code>__env.js</code> inside
+      thinkful-ui.</p>
+      <h3><Link to="/demo">
+        Interactive Component Demos
+      </Link></h3>
```
Comment (learykara) [inline client/HomePage.jsx]: I haven't seen this style yet with two tags opened/closed on the same lines.. is that preferable in some cases?

**Comment by learykara** [comment]
looks good to me!

File: client/routes.jsx (hunk 3 above)
Comment (tholex) [inline client/routes.jsx]: I was about to make a subroute with a modal -- will add it shortly..

**Comment by tholex** [comment]
@learykara updated w/ static modal on a  route

File: src/Modal/Modal.jsx [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+ * to show it conditionally on a user action.
+ */
+class Modal extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      isOpen: true,
+      controlledByParent: !! props.close
+    }
+  }
```
Comment (ripleyaffect) [inline src/Modal/Modal.jsx]: here is where you bind functions. something like
```es6
this._myFunction = this._myFunction.bind(this);
```

**Comment by ripleyaffect** [comment]
:+1:
